src/odm_sharing/private/trees.py

```diff
@@ -0,0 +1,408 @@
+
+from functional import seq
+
+from odm_sharing.private.common import TableName
+from odm_sharing.private.rules import ParseError, Rule, RuleId, RuleMode
+from odm_sharing.private.stdext import StrEnum
+from odm_sharing.private.utils import not_empty, qt
+
+
+# {{{1 types
+
+
+RuleTree = object  # opaque
+
+
+class Op(StrEnum):
+    AND = 'and'
+    EQ = '='
+    GT = '>'
+    GTE = '>='
+    LT = '<'
+    LTE = '<='
+    OR = 'or'
+    RANGE = 'in'
+
+
+class NodeKind(StrEnum):
+    ROOT = 'root'
+    SHARE = 'share'
+    TABLE = 'table'
+    SELECT = 'select'
+    GROUP = 'group'
+    FILTER = 'filter'
+    KEY = 'key'
+    VALUE = 'value'
+
+
+@dataclass(frozen=True)
+class Node:
+    rule_id: RuleId
+    kind: NodeKind
+    value: str = field(default_factory=str)
+    sons: list = field(default_factory=list)
+
+    @staticmethod
+    def _get_repr(node, depth: int = 0) -> str:  # type: ignore
+        result = ('    ' * depth) + str(node) + '\n'
+        for child in node.sons:
+            result += Node._get_repr(child, depth+1)
+        return result
+
+    def __repr__(self) -> str:
+        return Node._get_repr(self)
+
+    def __str__(self) -> str:
+        return f'({self.rule_id}, {self.kind}, {qt(self.value)})'
+
+
+class Ctx:
+    filename: str  # filename reference for error messages
+    node_tables: Dict[RuleId, List[str]]  # tables for each node
+    table_nodes: Dict[TableName, Dict[RuleId, Node]]  # node for each table
+    nodes: Dict[RuleId, Node]  # collection of nodes
+    root: Optional[Node]  # current tree root
+    rule_id: RuleId  # current node's rule-id
+
+    def __init__(self, filename: str) -> None:
+        self.filename = filename
+        self.node_tables = defaultdict(list)
+        self.table_nodes = defaultdict(dict)
+        self.nodes = {}
+        self.root = None
+        self.rule_id = 0
+
+
+# {{{1 constants
+
+
+ALL_LIT = "all"
+TABLE_MODES = [RuleMode.SELECT, RuleMode.FILTER]
+VAL_SEP = ";"
+
+
+# {{{1 error gen
+
+
+def gen_error(ctx: Ctx, desc: str) -> ParseError:
+    msg = f'{ctx.filename}(rule {ctx.rule_id}): {desc}'
+    print('Error: ' + msg, file=sys.stderr)
+    return ParseError(msg)
+
+
+def fail(ctx: Ctx, desc: str) -> None:
+    raise gen_error(ctx, desc)
+
+
+# {{{1 input text parsing
+
+
+def parse_list(ctx: Ctx, val_str: str,
+               min: int = 0, max: int = 0) -> List[str]:
+    '''splits a multiple-value string into a list
+
+    :param min: min required number of elements, or zero
+    :param max: max required number of elements, or zero
+    :raises ParseError:
+    '''
+    (a, b) = (min, max)
+    result = seq(val_str.split(VAL_SEP))\
+        .map(str.strip)\
+        .filter(not_empty)\
+        .list()
+    n = len(result)
+    if a == b:
+        if a != 0 and n != a:
+            fail(ctx, f'expected {a} values, got {n}')
+    elif b == 0:
+        if n < a:
+            fail(ctx, f'expected at least {a} values, got {n}')
+    elif n not in range(a, b-1):
```

Comment:
> Should this be `b + 1` since range returns one less than the max?

good catch, thanks.

> Reason for using range as opposed to just checking `n < a || n > b`?

you're right, it was a bad idea to use range. Fixed.